File: java/src/main/java/dev/mudbase/showcase/ecommerce/service/CartService.java
package dev.mudbase.showcase.ecommerce.service;

import dev.mudbase.showcase.ecommerce.auth.AuthSession;
import dev.mudbase.showcase.ecommerce.auth.GuestCartHolder;
import dev.mudbase.showcase.ecommerce.config.MudbaseProperties;
import dev.mudbase.showcase.ecommerce.domain.Cart;
import dev.mudbase.showcase.ecommerce.domain.CartItem;
import dev.mudbase.showcase.ecommerce.mudbase.JsonFields;
import dev.mudbase.showcase.ecommerce.mudbase.MudbaseDataClient;
import jakarta.servlet.http.HttpSession;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import org.springframework.stereotype.Service;

/**
 * Reads and writes the shopping cart: a plain HttpSession-held list before sign-in (Mudbase's
 * `customer`-role-only ownership condition on the `carts` collection denies an anonymous-guest
 * session entirely - see GuestCartHolder), and the real `carts` document, read-then-create-or-update
 * since there is no native upsert, once the shopper is signed in as a customer.
 */
@Service
public class CartService {

  private final MudbaseDataClient dataClient;
  private final MudbaseProperties properties;

  public CartService(MudbaseDataClient dataClient, MudbaseProperties properties) {
    this.dataClient = dataClient;
    this.properties = properties;
  }

  private boolean usesServerCart(AuthSession auth) {
    return auth != null && auth.isSignedIn() && auth.isCustomer();
  }

  private Optional<Cart> findServerCart(AuthSession auth) {
    List<Map<String, Object>> docs =
        dataClient.list(
            auth.getToken(),
            properties.getCartsCollectionId(),
            Map.of("userId", auth.getUserId()),
            "-createdAt",
            1,
            1);
    return docs.isEmpty() ? Optional.empty() : Optional.of(Cart.fromDocument(docs.get(0)));
  }

  public List<CartItem> getItems(HttpSession session, AuthSession auth) {
    if (usesServerCart(auth)) {
      return findServerCart(auth).map(Cart::getItems).orElseGet(List::of);
    }
    return GuestCartHolder.get(session);
  }

  public long getSubtotalCents(List<CartItem> items) {
    return items.stream().mapToLong(CartItem::getLineTotalCents).sum();
  }

  public void addItem(HttpSession session, AuthSession auth, CartItem item, int quantity) {
    List<CartItem> current = new ArrayList<>(getItems(session, auth));
    int existingIndex = indexOf(current, item.productId());
    if (existingIndex >= 0) {
      CartItem existing = current.get(existingIndex);
      current.set(existingIndex, existing.withQuantity(existing.quantity() + quantity));
    } else {
      current.add(item.withQuantity(quantity));
    }
    persist(session, auth, current);
  }

  public void updateQuantity(HttpSession session, AuthSession auth, String productId, int quantity) {
    List<CartItem> current = new ArrayList<>(getItems(session, auth));
    if (quantity <= 0) {
      current.removeIf(i -> i.productId().equals(productId));
    } else {
      int index = indexOf(current, productId);
      if (index >= 0) {
        current.set(index, current.get(index).withQuantity(quantity));
      }
    }
    persist(session, auth, current);
  }

  public void removeItem(HttpSession session, AuthSession auth, String productId) {
    List<CartItem> current = new ArrayList<>(getItems(session, auth));
    current.removeIf(i -> i.productId().equals(productId));
    persist(session, auth, current);
  }

  public void clear(HttpSession session, AuthSession auth) {
    persist(session, auth, List.of());
  }

  private int indexOf(List<CartItem> items, String productId) {
    for (int i = 0; i < items.size(); i++) {
      if (items.get(i).productId().equals(productId)) {
        return i;
      }
    }
    return -1;
  }

  private void persist(HttpSession session, AuthSession auth, List<CartItem> items) {
    if (!usesServerCart(auth)) {
      GuestCartHolder.set(session, items);
      return;
    }
    // Re-fetch right before deciding create-vs-update rather than trusting any caller-held
    // snapshot, so a cart created moments earlier by migrateGuestCartToServer() in the same
    // request sequence is never silently overwritten by a second, duplicate cart document.
    Optional<Cart> existing = findServerCart(auth);
    Map<String, Object> body = Map.of("itemsJson", JsonFields.stringify(items));
    if (existing.isPresent()) {
      dataClient.update(auth.getToken(), properties.getCartsCollectionId(), existing.get().getId(), body);
    } else {
      Map<String, Object> createBody = new LinkedHashMap<>(body);
      createBody.put("userId", auth.getUserId());
      dataClient.create(auth.getToken(), properties.getCartsCollectionId(), createBody);
    }
  }

  /** Called once, right after a guest completes registration or signs in at checkout. */
  public void migrateGuestCartToServer(HttpSession session, AuthSession auth) {
    List<CartItem> guestItems = GuestCartHolder.get(session);
    if (guestItems.isEmpty()) {
      return;
    }
    Optional<Cart> existing = findServerCart(auth);
    if (existing.isEmpty()) {
      dataClient.create(
          auth.getToken(),
          properties.getCartsCollectionId(),
          Map.of("userId", auth.getUserId(), "itemsJson", JsonFields.stringify(guestItems)));
      GuestCartHolder.clear(session);
      return;
    }
    List<CartItem> merged = new ArrayList<>(existing.get().getItems());
    for (CartItem guestItem : guestItems) {
      int index = indexOf(merged, guestItem.productId());
      if (index >= 0) {
        CartItem current = merged.get(index);
        merged.set(index, current.withQuantity(current.quantity() + guestItem.quantity()));
      } else {
        merged.add(guestItem);
      }
    }
    dataClient.update(
        auth.getToken(),
        properties.getCartsCollectionId(),
        existing.get().getId(),
        Map.of("itemsJson", JsonFields.stringify(merged)));
    GuestCartHolder.clear(session);
  }
}

File: csharp/MudbaseShowcase.Ecommerce/Services/CartService.cs
using Microsoft.Extensions.Options;
using MudbaseShowcase.Ecommerce.Models;
using MudbaseShowcase.Ecommerce.Options;

namespace MudbaseShowcase.Ecommerce.Services;

/// <summary>
/// Cart logic, direct port of web/src/hooks/useCart.ts's server/guest split.
///
/// A "customer" application role, not an anonymous or unauthenticated session, is required to
/// persist a cart in the `carts` collection — Mudbase's ownership-conditioned permission only
/// grants create/read/update/delete to the `customer` customRole. A guest session (anonymous JWT,
/// no customRole) is denied. This keeps the cart in the browser's server-side session state until
/// the shopper has a real customer account, then hands off to the server — see
/// <see cref="MigrateGuestCartToServerAsync"/>, called right after registration/login succeeds in
/// the checkout flow.
/// </summary>
public sealed class CartService
{
    private readonly MudbaseDataService _data;
    private readonly MudbaseSessionAccessor _session;
    private readonly MudbaseOptions _options;

    public CartService(MudbaseDataService data, MudbaseSessionAccessor session, IOptions<MudbaseOptions> options)
    {
        _data = data;
        _session = session;
        _options = options.Value;
    }

    public async Task<IReadOnlyList<CartItem>> GetItemsAsync(CancellationToken cancellationToken)
    {
        MudbaseSessionUser? user = _session.CurrentUser;
        if (user is { IsCustomer: true })
        {
            CartDocument? cart = await GetCustomerCartAsync(user.Id, cancellationToken);
            return cart is null ? Array.Empty<CartItem>() : JsonFieldHelper.Parse(cart.ItemsJson, new List<CartItem>());
        }

        return ReadGuestCart();
    }

    public static int SubtotalCents(IReadOnlyList<CartItem> items) => items.Sum(i => i.LineTotalCents);

    public async Task AddItemAsync(CartItem newItem, int quantity, CancellationToken cancellationToken)
    {
        List<CartItem> items = (await GetItemsAsync(cancellationToken)).ToList();
        int existingIndex = items.FindIndex(i => i.ProductId == newItem.ProductId);

        if (existingIndex >= 0)
        {
            CartItem existing = items[existingIndex];
            items[existingIndex] = new CartItem
            {
                ProductId = existing.ProductId,
                Name = existing.Name,
                PriceCents = existing.PriceCents,
                Currency = existing.Currency,
                ImageUrl = existing.ImageUrl,
                Quantity = existing.Quantity + quantity,
            };
        }
        else
        {
            items.Add(new CartItem
            {
                ProductId = newItem.ProductId,
                Name = newItem.Name,
                PriceCents = newItem.PriceCents,
                Currency = newItem.Currency,
                ImageUrl = newItem.ImageUrl,
                Quantity = quantity,
            });
        }

        await PersistAsync(items, cancellationToken);
    }

    public async Task UpdateQuantityAsync(string productId, int quantity, CancellationToken cancellationToken)
    {
        List<CartItem> items = (await GetItemsAsync(cancellationToken)).ToList();
        items = quantity <= 0
            ? items.Where(i => i.ProductId != productId).ToList()
            : items.Select(i => i.ProductId == productId ? WithQuantity(i, quantity) : i).ToList();

        await PersistAsync(items, cancellationToken);
    }

    public async Task RemoveItemAsync(string productId, CancellationToken cancellationToken)
    {
        List<CartItem> items = (await GetItemsAsync(cancellationToken)).Where(i => i.ProductId != productId).ToList();
        await PersistAsync(items, cancellationToken);
    }

    public async Task ClearAsync(CancellationToken cancellationToken)
    {
        await PersistAsync(new List<CartItem>(), cancellationToken);
    }

    /// <summary>Called once, right after a guest completes real customer registration or signs in at checkout.</summary>
    public async Task MigrateGuestCartToServerAsync(string userId, CancellationToken cancellationToken)
    {
        List<CartItem> guestItems = ReadGuestCart();
        if (guestItems.Count == 0)
        {
            return;
        }

        CartDocument? existing = await GetCustomerCartAsync(userId, cancellationToken);

        if (existing is null)
        {
            await _data.CreateAsync<CartDocument>(_options.CartsCollectionId, new Dictionary<string, object?>
            {
                ["userId"] = userId,
                ["itemsJson"] = JsonFieldHelper.Stringify(guestItems),
            }, cancellationToken);
            _session.ClearGuestCart();
            return;
        }

        List<CartItem> merged = JsonFieldHelper.Parse(existing.ItemsJson, new List<CartItem>());
        foreach (CartItem guestItem in guestItems)
        {
            int idx = merged.FindIndex(i => i.ProductId == guestItem.ProductId);
            merged = idx >= 0
                ? merged.Select((i, index) => index == idx ? WithQuantity(i, i.Quantity + guestItem.Quantity) : i).ToList()
                : merged.Append(guestItem).ToList();
        }

        await _data.UpdateAsync<CartDocument>(_options.CartsCollectionId, existing.Id, new Dictionary<string, object?>
        {
            ["itemsJson"] = JsonFieldHelper.Stringify(merged),
        }, cancellationToken);

        _session.ClearGuestCart();
    }

    private static CartItem WithQuantity(CartItem item, int quantity) => new()
    {
        ProductId = item.ProductId,
        Name = item.Name,
        PriceCents = item.PriceCents,
        Currency = item.Currency,
        ImageUrl = item.ImageUrl,
        Quantity = quantity,
    };

    private List<CartItem> ReadGuestCart() => JsonFieldHelper.Parse(_session.GetGuestCartJson(), new List<CartItem>());

    private void WriteGuestCart(List<CartItem> items) => _session.SetGuestCartJson(JsonFieldHelper.Stringify(items));

    private async Task<CartDocument?> GetCustomerCartAsync(string userId, CancellationToken cancellationToken)
    {
        MudbaseListResult<CartDocument> result = await _data.ListAsync<CartDocument>(
            _options.CartsCollectionId, new Dictionary<string, object?> { ["userId"] = userId }, limit: 1, cancellationToken: cancellationToken);
        return result.Items.FirstOrDefault();
    }

    private async Task PersistAsync(List<CartItem> items, CancellationToken cancellationToken)
    {
        MudbaseSessionUser? user = _session.CurrentUser;
        if (user is not { IsCustomer: true })
        {
            WriteGuestCart(items);
            return;
        }

        string itemsJson = JsonFieldHelper.Stringify(items);

        // Re-fetch the authoritative cart right before deciding create-vs-update rather than
        // trusting any caller-held reference — a returning customer may already have a server cart
        // from an earlier session, and Mudbase has no native upsert.
        CartDocument? existing = await GetCustomerCartAsync(user.Id, cancellationToken);

        if (existing is not null)
        {
            await _data.UpdateAsync<CartDocument>(_options.CartsCollectionId, existing.Id, new Dictionary<string, object?>
            {
                ["itemsJson"] = itemsJson,
            }, cancellationToken);
        }
        else
        {
            await _data.CreateAsync<CartDocument>(_options.CartsCollectionId, new Dictionary<string, object?>
            {
                ["userId"] = user.Id,
                ["itemsJson"] = itemsJson,
            }, cancellationToken);
        }
    }
}
